### PR TITLE
fix: client.add not adding buffers properly

### PR DESCRIPTION
**Pull request by JohnTJohnston**
Fixed while getting AD thumbnailPhoto to work.

To repro, add a thumbnailPhoto to AD.

Note: I don't believe I ever reported this problem, but I've been using the fix for a long time.

### CODE REVIEW

**Review by UziTech — CHANGES_REQUESTED** [review]
Can you add a test?

**Comment by JohnTJohnston** [comment]
Sorry, I must have missed something in your documentation.  I don't know what is assumed by your test runs.  I need to be able to add a thumbnail photo to AD.  Do you run your tests against AD?  Can I count on the existence of a particular user?

File: lib/client/client.js [hunk 1]
```diff
@@ -238,6 +238,8 @@ Client.prototype.add = function add (name, entry, controls, callback) {
         save[k].forEach(function (v) {
           attr.addValue(v.toString())
         })
+      } else if (Buffer.isBuffer(save[k])) {
+        attr.addValue(save[k]);
```
Comment (jsumners) [inline lib/client/client.js:242]: Please follow the linting rules. Use `npm run lint`.

**Review by jsumners — CHANGES_REQUESTED** [review]
> To repro, add a thumbnailPhoto to AD.

Sorry. This is not an effective means of proving the change. First, not everyone has access to an Active Directory instance. Second, it does not prevent regressions in future releases. A unit test is required to satisfy these requirements.

**Comment by JohnTJohnston** [comment]
Sorry, I'm not clear what is expected of your unit tests.  How are they supposed to work without a back end directory to verify functionality?  Not sure what other attributes require passing a buffer into add, so how to test?

Seems like I may have to live with my private patches...

**Comment by jsumners** [comment]
> Sorry, I'm not clear what is expected of your unit tests. How are they supposed to work without a back end directory to verify functionality? Not sure what other attributes require passing a buffer into add, so how to test?
> 
> Seems like I may have to live with my private patches...

The client unit tests are in https://github.com/ldapjs/node-ldapjs/blob/91de94d1a625b66b753091697632cbc6f433913d/test/client.test.js

We further have "integration" tests at https://github.com/ldapjs/node-ldapjs/tree/91de94d1a625b66b753091697632cbc6f433913d/test-integration/client

The integration tests use an standard OpenLDAP instance as built by https://github.com/ldapjs/docker-test-openldap and provided via https://github.com/ldapjs/docker-test-openldap/pkgs/container/docker-test-openldap/openldap

The integration tests are useful in cases where writing a unit test is either too difficult or not sufficient.

Note: I need to do some work to update the Docker repository to GitHub's latest, truly anonymous, access mechanisms. But it is attainable currently through authenticated means.

File: lib/client/client.js (hunk 1 above)
Comment (JohnTJohnston) [reply lib/client/client.js]: npm run lint completed

note: I have eslint installed globally, but it was not working.  So I installed it as a local dev dependency along with required peers pulled in by extends standard: eslint-plugin-import eslint-plugin-node, and eslint-plugin-promise.

Please advise as to best practices here.  Seems like these should be included as dev dependencies in the repo.

**Comment by JohnTJohnston** [comment]
I am not familiar with openldap attributes.  this bug fix is specfic to attributes that require a javascript buffer to be passed in, and for all I know, openldap might have been smart enough not to have such attributes.

File: lib/client/client.js (hunk 1 above)
Comment (jsumners) [reply lib/client/client.js]: `npm install` should install all required dev dependencies. I know we were using vanilla `standard` but that switched to support some configuration extension. Maybe we missed adding some dependencies @UziTech ?

**Comment by jsumners** [comment]
> I am not familiar with openldap attributes. this bug fix is specfic to attributes that require a javascript buffer to be passed in, and for all I know, openldap might have been smart enough not to have such attributes.

We really can't merge patches without a test to cover their functionality. This PR addresses reading of binary attributes, yes? My initial theory would be that updating the server embedded in the unit tests to return an entry that has a `Buffer` valued attribute would allow for a unit test to be written to cover this change.

File: lib/client/client.js (hunk 1 above)
Comment (JohnTJohnston) [reply lib/client/client.js]: No, it's there.  Not sure what my problem was.  Sorry for this noise.

**Comment by JohnTJohnston** [comment]
This was necessary to modify a binary attribute.  This following test passes, but it passes both before and after the fix.

```
tap.test('add success with buffer', function (t) {
  const entry = {
    someAttribute: Buffer.alloc(20)
  }
  t.context.client.add('cn=add, ' + SUFFIX, entry, function (err, res) {
    t.error(err)
    t.ok(res)
    t.equal(res.status, 0)
    t.end()
  })
})
```

**Comment by jsumners** [comment]
> This following test passes, but it passes both before and after the fix.

Then we should investigate why.

**Review by UziTech — APPROVED** [review]

**Review by jsumners — APPROVED** [review]
Awesome! This is what we need to help improve this library. The test fails before the patch and succeeds after.

**Comment by jsumners** [comment]
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.

**Pull request merged**